api4/shared_channel.go

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package api4
+
+import (
+	"encoding/json"
+	"net/http"
+)
+
+func (api *API) InitSharedChannels() {
+	api.BaseRoutes.SharedChannels.Handle("/getremote/{remote_id:[A-Za-z0-9]+}", api.ApiSessionRequired(getRemoteClusterById)).Methods("GET")
+}
+
+func getRemoteClusterById(c *Context, w http.ResponseWriter, r *http.Request) {
+	// make sure remote cluster service is enabled.
+	if _, appErr := c.App.GetRemoteClusterService(); appErr != nil {
+		c.Err = appErr
+		return
+	}
+
+	c.RequireRemoteId()
+	if c.Err != nil {
+		return
+	}
+
+	// GetRemoteClusterForUser will only return a remote if the user is a member of at
+	// least one channel shared by the remote. All other cases return error.
+	rc, err := c.App.GetRemoteClusterForUser(c.Params.RemoteId, c.App.Session().UserId)
```

Comment:
This call can return both an error due to noRowsFound but also a "canonical" error. What's the reason that we are returning 404 always here? I would have expected a 404 on the former case and a 500 on the latter.

Reply:
Main reason was I was not using AppError in the app layer.  AppError is the biggest source of crashes in the entire app and I wish we'd get rid of it.  Anyway, I made this API use AppError and it now differentiates between 500 and 404.  Fixed via https://github.com/mattermost/mattermost-server/pull/17063/commits/9288e226865aae981b33cc140fe383c4028505c8